src/analysis/retail/warlock/shared/talents/DarkPact.tsx

```diff
@@ -0,0 +1,163 @@
+import { formatNumber, formatPercentage } from 'common/format';
+import TALENTS from 'common/TALENTS/warlock';
+import Analyzer, { Options, SELECTED_PLAYER } from 'parser/core/Analyzer';
+import Events, {
+  AbsorbedEvent,
+  ApplyBuffEvent,
+  CastEvent,
+  RemoveBuffEvent,
+} from 'parser/core/Events';
+import BoringSpellValueText from 'parser/ui/BoringSpellValueText';
+import Statistic from 'parser/ui/Statistic';
+import STATISTIC_CATEGORY from 'parser/ui/STATISTIC_CATEGORY';
+import STATISTIC_ORDER from 'parser/ui/STATISTIC_ORDER';
+
+const debug = true;
+
+interface DarkPactCast {
+  applyEvent: ApplyBuffEvent;
+  removeEvent?: RemoveBuffEvent;
+  castEvent?: CastEvent;
+  hpPostCast?: number;
+  hpPercentPreCast?: number;
+  maxHpPostCast?: number;
+  totalAbsorb?: number;
+  amountAbsorbed: number;
+  unusedAbsorb?: number;
+}
+
+// TODO: needs guide, defensives section, this can and should be used as often as possible to help healers out
+// Expected event order is ApplyBuff > CastEvent > Absorbs > RemoveBuff
+class DarkPact extends Analyzer {
+  casts: DarkPactCast[] = [];
+  darkPactActive = false;
+  hasIchor: boolean;
+
+  constructor(options: Options) {
+    super(options);
+    this.active = this.selectedCombatant.hasTalent(TALENTS.DARK_PACT_TALENT);
+    this.hasIchor = this.selectedCombatant.hasTalent(TALENTS.ICHOR_OF_DEVILS_TALENT);
+
+    if (!this.active) {
+      return;
+    }
+
+    this.addEventListener(
+      Events.cast.by(SELECTED_PLAYER).spell(TALENTS.DARK_PACT_TALENT),
+      this.onCast,
+    );
+    this.addEventListener(
+      Events.applybuff.to(SELECTED_PLAYER).spell(TALENTS.DARK_PACT_TALENT),
+      this.onApplyShield,
+    );
+    this.addEventListener(
+      Events.removebuff.to(SELECTED_PLAYER).spell(TALENTS.DARK_PACT_TALENT),
+      this.onRemoveShield,
+    );
+    this.addEventListener(
+      Events.absorbed.by(SELECTED_PLAYER).spell(TALENTS.DARK_PACT_TALENT),
+      this.onAbsorb,
+    );
+  }
+
+  onCast(event: CastEvent) {
+    if (!this.darkPactActive) {
+      this.warn('Unexpected Dark Pact CastEvent, no ApplyBuffEvent before?');
+      return;
+    }
+    debug &&
+      (!event.hitPoints || !event.maxHitPoints) &&
+      this.debug('Dark Pact CastEvent has no HP data');
+
+    this.casts[this.casts.length - 1].castEvent = event;
+    this.casts[this.casts.length - 1].hpPostCast = event.hitPoints;
+    this.casts[this.casts.length - 1].maxHpPostCast = event.maxHitPoints;
+    this.casts[this.casts.length - 1].hpPercentPreCast =
```

Comment:
minor nit: this could be simplified:

```suggestion
    const lastCast = this.casts.at(-1)!; // we know that lastCast is defined because `this.darkPactActive` is true
    lastCast.castEvent = event;
    lastCast.hpPostCast = event.hitPoints;
    lastCast.maxHpPostCast = event.maxHitPoints;
    lastCast.hpPercentPreCast =
```

very minor tho